hazelcast/src/main/java/com/hazelcast/query/impl/bitmap/BitmapAlgorithms.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl.bitmap;
+
+/**
+ * Provides algorithms crucial for set operations on ordered iterators provided
+ * by sparse bit sets.
+ */
+final class BitmapAlgorithms {
+
+    private BitmapAlgorithms() {
+    }
+
```

Comment:
We can introduce one more algorithm for `AndNotIterator` which will try to evaluate a combination of `and`, and `not` operators in one pass. I expect that normally `not` operator will be used in combination with 'and'.
Lets say you have a predicate `A and not B`. Then we have 2 iterators A, B. Imagine a min-heap priority queue which returns an iterator with the minimum current index. We will use it to make an evaluation. 
First, the algorithm adds iterators A, B to the priority queue. An iterator in addition to the current index keeps track whether the value is required or prohibited. 
With the priority queue we find a match so that all iterators have the same position. Then we check with the prohibited mask that the match doesn't contain `not` element and that is how we find a final match. 
The main advantage of the approach that we don't have to make a real negate operator which is normally very expensive. The same idea applies to the `A or not B` predicates.

Reply:
I really like the idea, but we have no time to implement it. A potential subject for future TDD(s).